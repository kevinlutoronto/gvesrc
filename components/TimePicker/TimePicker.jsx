// @flow
import React from 'react';
import cn from 'classnames';
import ReactSelect from 'react-select';
import { omit } from 'lodash';
import type { GlobalAttributes, SyntheticFocusEventHandler } from '../../types';

import styles from './style.scss';

/**
 * A TimePicker component.
 *
 * Optional 'value' must be in 24H 'HH:mm' format.
 *
 * Optional 'minTime' in 24H 'HH:mm' format, default time 00:00.
 *
 * Optional 'maxTime' in 24H 'HH:mm' format, default time 24:00.
 *
 * Optional 'timeIntervalMinutes', Time step for time slots, default is 30 minute.
 *
 * Optional 'timeFormatType' label type in 24H or 12H, default 12H.
 *
 * Optional 'defaultOption':
 *    nextInterval = will round up: 11:26 -> 11:30  based on 'timeIntervalMinutes'.
 *    minimum (default selection) =  will select 'minTime'.
 *
 * Optional 'menuIsOpen': Whether dropdown is open. Default: false.
 *
 * Optional 'disabled'.
 *
 * Optional 'error'.
 *
 * Optional 'required'.
 *
 * Optional 'name'.
 *
 * Optional 'id'.
 *
 * Optional 'tabIndex'.
 *
 * onChange receives string value in 24H 'HH:mm' format.
 *
 * Optional 'onBlur'.
 *
 * Optional 'onFocus'.
 */

const FormatTypes = Object.freeze({
  timeFormat12: new Intl.DateTimeFormat('default', {
    hour: '2-digit',
    minute: '2-digit',
    hour12: true,
  }),
  timeFormat24: new Intl.DateTimeFormat('default', {
    hour: '2-digit',
    minute: '2-digit',
    hour12: false,
  }),
});

export type TimeFormatType = $Values<typeof FormatTypes>;

const DefaultOptions = Object.freeze({
  minimum: 'minimum',
  nextInterval: 'nextInterval',
});

export type DefaultOption = $Values<typeof DefaultOptions>;

type TimeOption = {|
  label: string,
  value: string,
|};

type Props = {|
  maxTime?: string,
  minTime?: string,
  defaultOption?: DefaultOption,
  timeFormat: TimeFormatType,
  timeIntervalMinutes?: number,
  value?: string,
  borderless?: boolean,
  className?: string,
  disabled?: boolean,
  error?: boolean,
  menuIsOpen?: boolean,
  onBlur?: (?string) => void,
  onChange?: string => void,
  onFocus?: SyntheticFocusEventHandler<>,
  placeholder?: string,
  name?: string,
  required?: boolean,
  ...GlobalAttributes,
|};

const isBefore = (date: Date, otherDate: Date) => date.getTime() < otherDate.getTime();
const isAfter = (date: Date, otherDate: Date) => date.getTime() > otherDate.getTime();
// using an arbitrary date as we only care about time
const get24HourTimeAsDate = (value?: string): Date => new Date(`2018-01-01 ${value || '00:00'}`);

const enumerateOptions = (
  start: string,
  end: string,
  intervalMinutes: number,
  timeFormat: Intl$DateTimeFormat
): Array<TimeOption> => {
  const minTime = get24HourTimeAsDate(start);
  const maxTime = get24HourTimeAsDate(end);

  const timeValue = new Date(minTime);
  const options = [];

  while (timeValue.getTime() <= maxTime.getTime()) {
    options.push({
      value: FormatTypes.timeFormat24.format(timeValue),
      label: timeFormat.format(timeValue),
    });

    timeValue.setMinutes(timeValue.getMinutes() + intervalMinutes);
  }

  return options;
};

const ClockIcon = () => <span className={styles.selectClockIcon} />;

class TimePicker extends React.Component<Props> {
  static DefaultOptions = DefaultOptions;

  static FormatTypes = FormatTypes;

  static defaultProps = {
    minTime: '00:00',
    maxTime: '24:00',
    timeFormat: FormatTypes.timeFormat24,
    timeIntervalMinutes: 30,
    placeholder: 'Select time',
  };

  getOptions = () =>
    enumerateOptions(
      this.props.minTime || TimePicker.defaultProps.minTime,
      this.props.maxTime || TimePicker.defaultProps.maxTime,
      this.props.timeIntervalMinutes || TimePicker.defaultProps.timeIntervalMinutes,
      this.props.timeFormat
    );

  getValue(value: ?string): ?string {
    if (!value) {
      return this.defaultTimeSelected();
    }

    const { maxTime, minTime } = this.props;

    const start = get24HourTimeAsDate(minTime);
    const end = get24HourTimeAsDate(maxTime);
    const valueAsDate = get24HourTimeAsDate(value);

    if (isBefore(valueAsDate, start)) {
      return minTime;
    }

    if (isAfter(valueAsDate, end)) {
      return maxTime;
    }

    return value;
  }

  handleChange = (option: TimeOption) => {
    const { onChange } = this.props;

    if (onChange) {
      onChange(option.value);
    }
  };

  handleOnBlur = (): void => {
    const { onBlur, value } = this.props;

    if (onBlur) {
      onBlur(value);
    }
  };

  defaultTimeSelected(): ?string {
    const {
      timeIntervalMinutes = TimePicker.defaultProps.timeIntervalMinutes,
      defaultOption,
    } = this.props;
    switch (defaultOption) {
      case TimePicker.DefaultOptions.nextInterval: {
        const now = new Date();
        const roundedUp = Math.ceil(now.getMinutes() / timeIntervalMinutes) * timeIntervalMinutes;

        const defaultTime = new Date(now);
        defaultTime.setMinutes(roundedUp);
        defaultTime.setSeconds(0);

        return TimePicker.FormatTypes.timeFormat24.format(defaultTime);
      }
      case TimePicker.DefaultOptions.minimum:
        return TimePicker.FormatTypes.timeFormat24.format(get24HourTimeAsDate(this.props.minTime));
      default:
        return null;
    }
  }

  render() {
    const {
      borderless,
      className,
      disabled,
      error,
      menuIsOpen,
      value,
      placeholder,
      ...rest
    } = this.props;

    const timeSelectClassName = cn(styles.selectInput, className, {
      [styles.borderless]: borderless,
      [styles.selectInputError]: error,
    });

    const currentValue = this.getValue(value);
    const options = this.getOptions();
    const currentOption = options.find(item => item.value === currentValue);

    const restProps = omit(
      rest,
      'defaultOption',
      'hidden',
      'maxTime',
      'minTime',
      'timeFormat',
      'timeIntervalMinutes',
      'title'
    );

    return (
      <ReactSelect
        {...restProps}
        menuIsOpen={menuIsOpen}
        components={{
          IndicatorSeparator: null,
          DropdownIndicator: ClockIcon,
        }}
        className={timeSelectClassName}
        classNamePrefix="time-picker"
        clearable={false}
        options={options}
        searchable={false}
        isDisabled={disabled}
        value={currentOption}
        onChange={this.handleChange}
        placeholder={placeholder}
        onBlur={this.handleOnBlur}
      />
    );
  }
}

TimePicker.displayName = 'Plasma@TimePicker';

export default TimePicker;
