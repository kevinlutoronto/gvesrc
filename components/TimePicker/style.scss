// TODO(grozki): Disabling this rule because it breaks with either solution tried. Should be attempted again later,
//  maybe a better solution exists.
// stylelint-disable no-descending-specificity
@import '../../styles/base';
@import '../../styles/colors';

// These are here just to shut ESLint up.
/* stylelint-disable block-no-empty */
.borderless {
}

.selectInput {
}

.selectInputError {
}
/* stylelint-enable block-no-empty */

.selectClockIcon {
  background: url('data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHdpZHRoPSIyMCIgaGVpZ2h0PSIyMCIgdmlld0JveD0iMCAwIDIwIDIwIj48cGF0aCBmaWxsPSIjQzFDMUMyIiBmaWxsLXJ1bGU9ImV2ZW5vZGQiIHN0cm9rZT0iI0MxQzFDMiIgc3Ryb2tlLXdpZHRoPSIuNSIgZD0iTTEwIDIuMTI1YTcuODc1IDcuODc1IDAgMSAwIDAgMTUuNzUgNy44NzUgNy44NzUgMCAwIDAgMC0xNS43NXpNMSAxMGE5IDkgMCAxIDEgMTggMCA5IDkgMCAwIDEtMTggMHptOSAwaDQuODc1djEuMTI0aC02di02SDEwVjEweiIvPjwvc3ZnPg==')
  transparent no-repeat center center;
  display: inline-block;
  width: 20px;
  height: 20px;
  padding: 5px;
}

:global {
  :local(.selectInput) {
    @extend %type--body;

    width: 7.5rem;
    font-size: 13px;
    color: $cBlack50;
    -webkit-font-smoothing: antialiased;

    .time-picker__single-value {
      color: currentColor;
    }

    .time-picker__placeholder {
      color: $cGray400;
    }

    .time-picker__indicators {
      order: 0;
    }

    .time-picker__value-container {
      order: 1;
      padding: 1px;
    }

    .time-picker__control {
      @extend %input--wrapper;

      &:hover {
        @extend %input--wrapper--hover;
      }
    }

    .time-picker__control.time-picker__control--is-focused {
      @extend %input--wrapper--focus;
    }

    .time-picker__control.time-picker__control--is-disabled {
      @extend %input--wrapper--disabled;

      background: white;
      pointer-events: auto;

      &:hover {
        @extend %input--wrapper--disabled;
      }

      .time-picker__placeholder {
        color: currentColor;
      }

      :local(.selectClockIcon) {
        opacity: 0.5;
      }
    }

    .time-picker__option {
      text-align: center;
      color: $cBlack50;

      &:active {
        background-color: $cGray400;
      }

      &--is-focused {
        background-color: $cGray600;
      }

      &--is-selected {
        background-color: $cGray600;
        color: currentColor;
      }
    }

    .time-picker__menu {
      margin-top: 0;
      border-radius: 0 0 2px 2px;
      box-shadow: 0 2px 4px rgba(0, 0, 0, 0.15);
    }
  }

  :local(.selectInput.selectInputError) {
    .time-picker__control {
      @extend %input--wrapper--error;

      &:hover {
        @extend %input--wrapper--error;
      }
    }

    .time-picker__control--is-focused {
      @extend %input--wrapper--error;
    }
  }

  :local(.selectInput.borderless) {
    border-color: transparent;
    box-shadow: none;

    .time-picker__control {
      border-color: transparent;

      &:hover {
        border-color: $cGray200;
        box-shadow: none;
      }

      &--is-focused {
        border-color: $cGray200;
      }
    }
  }
}
// stylelint-enable no-descending-specificity
