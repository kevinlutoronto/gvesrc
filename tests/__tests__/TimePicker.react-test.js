/* eslint-disable */

import React from 'react';
import renderer from 'react-test-renderer';
import TimePicker from '../../components/TimePicker/TimePicker';

describe('TimePicker default', () => {
  test('default', () => {
    const component = renderer.create(<TimePicker disabled />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('24H format', () => {
    const component = renderer.create(
      <TimePicker timeFormat={TimePicker.FormatTypes.timeFormat24} value="14:00" />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('12H format', () => {
    const component = renderer.create(
      <TimePicker timeFormat={TimePicker.FormatTypes.timeFormat12} value="14:00" />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
