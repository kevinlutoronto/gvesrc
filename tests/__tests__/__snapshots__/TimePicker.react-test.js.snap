// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimePicker default 12H format 1`] = `
<div
  className="css-10nd86i selectInput"
  id={undefined}
  onKeyDown={[Function]}
>
  <div
    className="css-vj8t7z time-picker__control"
    onMouseDown={[Function]}
    onTouchEnd={[Function]}
  >
    <div
      className="css-1hwfws3 time-picker__value-container time-picker__value-container--has-value"
    >
      <div
        className="css-xp4uvy time-picker__single-value"
      >
        2:00 PM
      </div>
      <div
        className="css-1g6gooi"
      >
        <div
          className="time-picker__input"
          style={
            Object {
              "display": "inline-block",
            }
          }
        >
          <input
            aria-autocomplete="list"
            aria-label={undefined}
            aria-labelledby={undefined}
            autoCapitalize="none"
            autoComplete="off"
            autoCorrect="off"
            className={undefined}
            disabled={false}
            id="react-select-4-input"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            spellCheck="false"
            style={
              Object {
                "background": 0,
                "border": 0,
                "boxSizing": "content-box",
                "color": "inherit",
                "fontSize": "inherit",
                "opacity": 1,
                "outline": 0,
                "padding": 0,
                "width": "1px",
              }
            }
            tabIndex="0"
            type="text"
            value=""
          />
          <div
            style={
              Object {
                "height": 0,
                "left": 0,
                "overflow": "scroll",
                "position": "absolute",
                "top": 0,
                "visibility": "hidden",
                "whiteSpace": "pre",
              }
            }
          >
            
          </div>
        </div>
      </div>
    </div>
    <div
      className="css-1wy0on6 time-picker__indicators"
    >
      <span
        className="selectClockIcon"
      />
    </div>
  </div>
</div>
`;

exports[`TimePicker default 24H format 1`] = `
<div
  className="css-10nd86i selectInput"
  id={undefined}
  onKeyDown={[Function]}
>
  <div
    className="css-vj8t7z time-picker__control"
    onMouseDown={[Function]}
    onTouchEnd={[Function]}
  >
    <div
      className="css-1hwfws3 time-picker__value-container time-picker__value-container--has-value"
    >
      <div
        className="css-xp4uvy time-picker__single-value"
      >
        14:00
      </div>
      <div
        className="css-1g6gooi"
      >
        <div
          className="time-picker__input"
          style={
            Object {
              "display": "inline-block",
            }
          }
        >
          <input
            aria-autocomplete="list"
            aria-label={undefined}
            aria-labelledby={undefined}
            autoCapitalize="none"
            autoComplete="off"
            autoCorrect="off"
            className={undefined}
            disabled={false}
            id="react-select-3-input"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            spellCheck="false"
            style={
              Object {
                "background": 0,
                "border": 0,
                "boxSizing": "content-box",
                "color": "inherit",
                "fontSize": "inherit",
                "opacity": 1,
                "outline": 0,
                "padding": 0,
                "width": "1px",
              }
            }
            tabIndex="0"
            type="text"
            value=""
          />
          <div
            style={
              Object {
                "height": 0,
                "left": 0,
                "overflow": "scroll",
                "position": "absolute",
                "top": 0,
                "visibility": "hidden",
                "whiteSpace": "pre",
              }
            }
          >
            
          </div>
        </div>
      </div>
    </div>
    <div
      className="css-1wy0on6 time-picker__indicators"
    >
      <span
        className="selectClockIcon"
      />
    </div>
  </div>
</div>
`;

exports[`TimePicker default default 1`] = `
<div
  className="css-1sontr1 selectInput time-picker--is-disabled"
  id={undefined}
  onKeyDown={[Function]}
>
  <div
    className="css-162g8z5 time-picker__control time-picker__control--is-disabled"
    onMouseDown={[Function]}
    onTouchEnd={[Function]}
  >
    <div
      className="css-1hwfws3 time-picker__value-container"
    >
      <div
        className="css-1492t68 time-picker__placeholder"
      >
        Select time
      </div>
      <div
        className="css-1g6gooi"
      >
        <div
          className="time-picker__input"
          style={
            Object {
              "display": "inline-block",
            }
          }
        >
          <input
            aria-autocomplete="list"
            aria-label={undefined}
            aria-labelledby={undefined}
            autoCapitalize="none"
            autoComplete="off"
            autoCorrect="off"
            className={undefined}
            disabled={true}
            id="react-select-2-input"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            spellCheck="false"
            style={
              Object {
                "background": 0,
                "border": 0,
                "boxSizing": "content-box",
                "color": "inherit",
                "fontSize": "inherit",
                "opacity": 1,
                "outline": 0,
                "padding": 0,
                "width": "1px",
              }
            }
            tabIndex="0"
            type="text"
            value=""
          />
          <div
            style={
              Object {
                "height": 0,
                "left": 0,
                "overflow": "scroll",
                "position": "absolute",
                "top": 0,
                "visibility": "hidden",
                "whiteSpace": "pre",
              }
            }
          >
            
          </div>
        </div>
      </div>
    </div>
    <div
      className="css-1wy0on6 time-picker__indicators"
    >
      <span
        className="selectClockIcon"
      />
    </div>
  </div>
</div>
`;
